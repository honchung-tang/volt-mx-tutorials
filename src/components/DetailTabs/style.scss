@import '../../assets/styles/kony';

.tabWrapper {
  width: 100%;

  div[role='tab'] {
    font-family: $RobotoSlabRegular;
    font-size: 16px;
    color: $secondary-text;
    padding: 20px 0;
    margin-right: 38px;

    &:hover {
      color: $secondary-text;
    }

    @include breakpoint(mobile) {
      color: $secondary-text-mobile;
      opacity: .5;

      &:hover {
        color: $secondary-text-mobile;
      }
    }
  }

  div[aria-selected='true'] {
    color: $title-text;

    &:hover {
      color: $title-text;
      cursor: default;
    }

    @include breakpoint(mobile) {
      color: $secondary-text-mobile-dark;
      opacity: 1;

      &:hover {
        color: $secondary-text-mobile-dark;
        opacity: 1;
      }
    }
  }
}
