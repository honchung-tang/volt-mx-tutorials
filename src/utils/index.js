import _startCase from 'lodash/startCase';
import _toLower from 'lodash/toLower';
import moment from 'moment';

/**
 *  Truncates the string.
 * @private
 * @param {string} string - The string to truncate.
 * @param {number} length - The length of truncated string.
 * @returns {string} Returns truncated `string`.
 */

export const truncate = (string, length) => {
  if (typeof length !== 'number' || typeof string !== 'string') return string;
  if (string.length <= length) return string;
  if (length <= 3) return `${string.slice(0, length)}...`;
  return `${string.slice(0, length - 3)}...`;
};

/**
 *  Truncates the string without cutting words.
 * @private
 * @param {string} string - The string to truncate.
 * @param {number} length - The length of truncated string.
 * @returns {string} Returns truncated `string`.
 */

export const truncateAdvance = (string, length) => {
  if (typeof length !== 'number' || typeof string !== 'string') return string;
  if (string.length <= length) return string;
  return `${string.replace(new RegExp(`^(.{${length}}[^\\s]*).*`), '$1')}`;
};

/**
 *  Capitalize First Letter of the string.
 * @private
 * @param {string} string - The string whose first letter to be capitalized.
 * @returns {string} Returns string whose first letter is capital.
 */

export const capitalizeFirstLetter = (string) => {
  if (!string) {
    return null;
  }
  return _startCase(_toLower(string));
};

/**
 *  Get month name from the timestamp.
 * @private
 * @param {number} timestamp - The timestamp which is going to converted.
 * @returns {string} Returns month name from timestamp passed.
 */

export const getMonthNameFromTimestamp = (timestamp) => {
  const date = new Date(timestamp);
  const monthName = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  return monthName[date.getMonth()];
};

/**
 *  Get device os platform.
 * @private
 * @returns {string} Returns name of platform.
 */

let os = null;
if (process.browser) {
  const { userAgent, platform } = window.navigator;
  const macosPlatforms = ['Macintosh', 'MacIntel', 'MacPPC', 'Mac68K'];
  const windowsPlatforms = ['Win32', 'Win64', 'Windows', 'WinCE'];
  const iosPlatforms = ['iPhone', 'iPad', 'iPod', 'iPhone Simulator', 'iPod Simulator', 'iPad Simulator'];
  if (macosPlatforms.indexOf(platform) !== -1) {
    os = 'Mac OS';
  } else if (iosPlatforms.indexOf(platform) !== -1) {
    os = 'iOS';
  } else if (windowsPlatforms.indexOf(platform) !== -1) {
    os = 'Windows';
  } else if (/Android/.test(userAgent)) {
    os = 'Android';
  } else if (!os && /Linux/.test(platform)) {
    os = 'Linux';
  }
}

export const getOS = os;

/**
 * Load a script.
 * @private
 * @param src of the file
 */

export const loadScript = (src) => {
  const tag = document.createElement('script');
  tag.src = src;
  tag.async = true;
  document.body.appendChild(tag);
};

export const validateData = (d) => {
  if (d) {
    if (typeof (d) === 'string' && d !== '') {
      return true;
    }
    if (typeof (d) === 'object' && Object.keys(d).length > 0) {
      return true;
    }
  }
  return false;
};

const escapeRegExp = string => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

/**
 * Replace all occurences of string with other.
 *
 * @private
 *
 * @param str, String to be replaced on.
 * @param term, term to be replaced.
 * @param replacement, String to be replaced with.
 */
export const replaceAll = (str, term, replacement) => str.replace(new RegExp(escapeRegExp(term), 'g'), replacement);

/**
 * Convert number into two digits.
 *
 * @private
 *
 * @param number, Number to be converted.
 *
 * @return String, Number in two digits.
 */
function numberToTwoDigits(number) {
  return number > 9 ? `${number}` : `0${number}`;
}

/**
 * Convert seconds into HH:mm:ss time format.
 *
 * @private
 *
 * @param second, Number of seconds.
 *
 * @return String, time in the HH:mm:ss format.
 */
export const secondsToTime = (second) => {
  const seconds = parseInt(second, 10);
  const momentDuration = moment.duration(seconds, 'seconds');
  const hours = numberToTwoDigits(Math.floor(momentDuration.asHours()));
  const mins = numberToTwoDigits(momentDuration.minutes());
  const secs = numberToTwoDigits(momentDuration.seconds());
  return `${hours}:${mins}:${secs}`;
};
