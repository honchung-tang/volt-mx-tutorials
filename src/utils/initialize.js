import Router from 'next/router';
import axios from 'axios';
import { Base64 } from 'js-base64';
import _size from 'lodash/size';
import * as actionType from '../actions/actionTypes';
import { marketplaceActions } from '../actions';
import { getCookie } from './cookies';
import { baseURL } from '../config/settings';

const tempInstance = axios.create({
  baseURL,
});

export default async function (ctx) {
  if (ctx.isServer) {
    if (ctx.query || ctx.req?.cookies) {
      if (ctx.query?.useragent || ctx.req?.cookies?.useragent) {
        const useragent = ctx.query?.useragent || ctx.req?.cookies?.useragent;
        if (useragent.toLowerCase() === 'visualizer') {
          ctx.store.dispatch(marketplaceActions.setVizWeb(true));
          ctx.res.cookie('useragent', 'visualizer');
          ctx.store.dispatch({
            type: actionType.SET_USER_ID,
            payload: ctx.req?.cookies.viz_user_id,
          });
          if (ctx.query.platform || ctx.req?.cookies.platform) {
            const platform = ctx.query.platform || ctx.req?.cookies.platform;
            ctx.store.dispatch(marketplaceActions.setVizVersion(platform));
            ctx.res.cookie('platform', platform);
          }
          if (ctx.query.nfiversion || ctx.req?.cookies.nfiversion) {
            const nfiversion = ctx.query.nfiversion || ctx.req?.cookies.nfiversion;
            ctx.store.dispatch(marketplaceActions.setVizVersion(nfiversion));
            ctx.res.cookie('nfiversion', nfiversion);
          }
        } else if (useragent.toLowerCase() === 'app-preview') {
          ctx.store.dispatch(marketplaceActions.setVizApp(true));
          ctx.res.cookie('useragent', 'app-preview');
        }
      }

      if (ctx.query?.usertype || ctx.req?.cookies?.usertype) {
        const usertype = ctx.query?.usertype || ctx.req?.cookies?.usertype;
        ctx.store.dispatch(marketplaceActions.setUserType(usertype));
        ctx.res.cookie('usertype', usertype);
      }

      if (ctx.query.cloud_account_id || ctx.req?.cookies?.account_id) {
        const accountId = ctx.query.cloud_account_id || ctx.req?.cookies?.account_id;
        ctx.res.cookie('account_id', accountId);
      }
    }

    // Set X-Kony-Authorization
    if (ctx.req?.cookies) {
      // Handle Authentication
      if (!ctx.req?.cookies.mp_user_id
        || !ctx.req?.cookies.mp_user_info
        || !ctx.req?.cookies.mp_user_data
        || !ctx.req?.cookies.mp_refresh_token) {
        ctx.res.cookie('mp_user_data', '', { maxAge: -1 });
        ctx.res.cookie('mp_refresh_token', '', { maxAge: -1 });
        ctx.res.cookie('mp_claims_token', '', { maxAge: -1 });
        ctx.res.cookie('mp_user_id', '', { maxAge: -1 });
        ctx.res.cookie('mp_user_info', '', { maxAge: -1 });
        ctx.res.cookie('mp_mod_menu', '', { maxAge: -1 });
        ctx.res.cookie('hike_admin_menu', '', { maxAge: -1 });
        ctx.res.cookie('mp_private_mod_menu', '', { maxAge: -1 });
        delete tempInstance.defaults.headers.common['X-Kony-Authorization'];
      } else {
        ctx.store.dispatch({
          type: actionType.SET_USER_ID,
          payload: ctx.req?.cookies.mp_user_id || null,
        });
        ctx.store.dispatch({
          type: actionType.SET_CLOUD_ACCOUNTS,
          payload: ctx.req?.cookies.mp_user_info ? JSON.parse(ctx.req?.cookies.mp_user_info) : [],
        });
        const userData = ctx.req?.cookies.mp_user_data
          ? JSON.parse(Base64.decode(ctx.req?.cookies.mp_user_data))
          : false;
        const auth_token = ctx.req?.cookies.mp_refresh_token || false;
        if (userData && auth_token) {
          tempInstance.defaults.headers.common['X-Kony-Authorization'] = auth_token;
          await ctx.store.dispatch({
            type: actionType.SET_AUTH,
            payload: { whoami: userData },
          });
          ctx.store.dispatch({
            type: actionType.SET_MP_MODERATOR_MENU,
            payload: ctx.req?.cookies.mp_mod_menu
              ? JSON.parse(ctx.req?.cookies.mp_mod_menu)
              : [],
          });
          ctx.store.dispatch({
            type: actionType.SET_HIKE_ADMIN_MENU,
            payload: ctx.req?.cookies.hike_admin_menu
              ? JSON.parse(ctx.req?.cookies.hike_admin_menu)
              : [],
          });
          ctx.store.dispatch({
            type: actionType.SET_PRIVATE_MODERATOR_MENU,
            payload: ctx.req?.cookies.mp_private_mod_menu
              ? JSON.parse(ctx.req?.cookies.mp_private_mod_menu)
              : [],
          });
          await ctx.store.dispatch(marketplaceActions.fetchBasecampUser(userData.primary_email));
        }
      }
    }

    if (ctx.query) {
      const {
        sortBy,
        cloud_account_id,
        category,
        platform,
        keyword,
        domain,
        channels,
        tags,
        author,
        cloudId,
        mpType,
        subdomain,
        nfiversion,
      } = ctx.query;
      let selectedFilters = {
        sortBy,
        domain,
        channels,
        category,
        platform,
        keyword,
        tags,
        mpType,
        author,
        cloudId,
        cloud_account_id,
        nfiversion,
      };

      let tempHeaders = {
        'X-KONY-SORTBY': ctx.query.sortBy || 'popular',
        'X-KONY-CATEGORY': ctx.query.category || 'all',
        'X-KONY-PLATFORM-VERSION': ctx.query.platform || 'all',
        'X-KONY-NFI-VERSION': ctx.query.nfiversion || 'all',
        'X-KONY-KEYWORD': ctx.query.keyword || '',
      };

      if (keyword && !sortBy) {
        delete tempHeaders['X-KONY-SORTBY'];
      }

      if (category) {
        selectedFilters.category = Array.isArray(category) ? category : category.split(',');
        tempHeaders['X-KONY-CATEGORY'] = selectedFilters.category.join('&');
      }

      if (tags) {
        selectedFilters.tags = Array.isArray(tags) ? tags : tags.split(',');
        tempHeaders['X-KONY-TAG'] = selectedFilters.tags.join('&');
      }

      if (domain) {
        selectedFilters.subdomain = subdomain;
        tempHeaders['X-KONY-DOMAIN'] = ctx.query.domain;
        if (subdomain) {
          tempHeaders['X-KONY-SUBDOMAIN'] = ctx.query.subdomain;
        }
      }

      if (channels) {
        selectedFilters.channels = Array.isArray(channels) ? channels : channels.split(',');
        tempHeaders['X-KONY-CHANNEL'] = selectedFilters.channels.join('&');
      }

      if (author) {
        try {
          const decodedAuthor = Base64.decode(author);
          if (decodedAuthor) {
            tempHeaders['X-KONY-AUTHOR'] = decodedAuthor.split('::')[1] || decodedAuthor.split('::')[0];
          }
        } catch (error) {
          tempHeaders['X-KONY-AUTHOR'] = author;
        }
      }

      if (cloudId) {
        const state = ctx.store.getState();
        if (!state.marketplace.isLogin) {
          ctx.res.redirect(`/oauth/login?destination=/cloud/${cloudId}`);
        }
        tempHeaders['X-KONY-ACCOUNT-ID'] = ctx.query.cloudId;
        tempHeaders['X-KONY-MARKETPLACE-TYPE'] = mpType || 'private';
      }

      if (ctx.query.cloud_account_id || ctx.req?.cookies?.account_id) {
        const accountId = ctx.query.cloud_account_id || ctx.req?.cookies?.account_id;
        ctx.res.cookie('account_id', accountId);
        tempHeaders['X-KONY-ACCOUNT-ID'] = accountId;

        if (ctx.query?.useragent || ctx.req?.cookies?.useragent) {
          tempHeaders['X-KONY-MARKETPLACE-TYPE'] = 'all';
        }
      }

      if (ctx.query?.useragent || ctx.req?.cookies?.useragent) {
        if (ctx.query.platform || ctx.req?.cookies.platform) {
          tempHeaders['X-KONY-PLATFORM-VERSION'] = ctx.query.platform || ctx.req?.cookies.platform || 2352;
        } else {
          tempHeaders['X-KONY-PLATFORM-VERSION'] = 2352;
        }
        if (ctx.query.nfiversion || ctx.req?.cookies.nfiversion) {
          tempHeaders['X-KONY-NFI-VERSION'] = ctx.query.nfiversion || ctx.req?.cookies.nfiversion || 'all';
        } else {
          tempHeaders['X-KONY-NFI-VERSION'] = 'all';
        }
      }

      if (mpType) {
        tempHeaders['X-KONY-MARKETPLACE-TYPE'] = mpType || 'all';
      }

      ctx.store.dispatch({
        type: 'SET_SELECTED_FILTERS',
        payload: selectedFilters,
      });
      ctx.store.dispatch({ type: 'SET_HEADERS', payload: tempHeaders });
    }
  } else {
    const login = ctx.store.getState();
    if (login && ctx.pathname === '/auth/login') {
      setTimeout(() => {
        Router.push('/');
      }, 0);
    }
  }
}

if (process.browser) {
  const whoamiData = getCookie('mp_user_data') ? JSON.parse(Base64.decode(getCookie('mp_user_data'))) : false;

  const auth_token = getCookie('mp_refresh_token') || false;
  if (whoamiData && auth_token) {
    tempInstance.defaults.headers.common['X-Kony-Authorization'] = auth_token;
  }
}

export const urlRedirect = (path, query) => {
  let base = '';
  let urlPath = path;
  if (process.browser) {
    if (getCookie('useragent') === 'visualizer') {
      base = '/marketplace-viz';
    }

    if (urlPath === '/r' && _size(query) === 0) {
      urlPath = '/';
    }
    Router.replace({
      pathname: `${base}${urlPath}`,
      query,
    });
  }
};

export const getUrl = (path, query, isVizWeb) => {
  let base = '';
  if (process.browser) {
    if (getCookie('useragent') === 'visualizer') {
      base = '/marketplace-viz';
    }
  } else if (isVizWeb) {
    base = '/marketplace-viz';
  }
  if (query && query.length > 0) {
    return `${base}${path}?${query}`;
  }
  return `${base}${path}`;
};
export const instance = tempInstance;
