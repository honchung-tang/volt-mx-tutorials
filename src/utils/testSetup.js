/* eslint-disable import/no-extraneous-dependencies */
require('jest-chain');
require('jest-extended');
