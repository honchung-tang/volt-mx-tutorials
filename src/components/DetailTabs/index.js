import React, { Component } from 'react';
import Tabs from 'antd/lib/tabs';
import Col from 'antd/lib/col';
import Row from 'antd/lib/row';
import PropTypes from 'prop-types';
import style from './style.scss';
import DetailsTabContent from '../DetailsTabContent';
import Documentation from '../Documentation';
import KonySupport from '../KonySupport';
import KonyLike from '../KonyLike';
import AssetDownloadButton from '../AssetDownloadButton';
import { connect } from 'react-redux';

const { TabPane } = Tabs;
const isDocLinkValid = src => src.length !== 0 && src !== '' && src !== null;

class DetailTabs extends Component {
  setKey = (key) => {
    const tempStore = JSON.parse(localStorage.getItem('selectedKeys')) || {};
    tempStore[this.props.ID] = key;
    localStorage.setItem('selectedKeys', JSON.stringify(tempStore));
  };

  renderHowToTab() {
    if (isDocLinkValid(this.props.DocumentationURL)) {
      return (
        <TabPane tab="How to" key="2"><Documentation isInIframe={this.props.Iframe} src={this.props.DocumentationURL} /></TabPane>
      );
    }
    return null;
  }

  render() {
    const {platformVersion, isVizWeb, termMapName} = this.props;
    return (
      <Tabs
        tabPosition="top"
        animated={false}
        defaultActiveKey="1"
        className={`${style.tabWrapper} detailPageTabWrapper`}
        onChange={this.setKey}
        tabBarExtraContent={(
          <Row className="nocontent">
            <Col md={0} lg={24} sm={0} xs={0}>
              <AssetDownloadButton
                {...this.props}
                title={this.props.title}
                url={this.props.url}
              />
            </Col>
            <KonyLike id={this.props.itemId} isLiked={this.props.isLiked} />
          </Row>
        )}
      >
        <TabPane tab="Details" key="1"><DetailsTabContent {...this.props} /></TabPane>
        {this.renderHowToTab()}
        <TabPane tab="Support" key="3"><KonySupport title={this.props.title} url={this.props.url} /></TabPane>
      </Tabs>
    );
  }
}

DetailTabs.propTypes = {
  asset: PropTypes.object,
  DocumentationURL: PropTypes.string,
};

DetailTabs.defaultProps = {
  asset: {},
  DocumentationURL: '',
};
export default DetailTabs;
