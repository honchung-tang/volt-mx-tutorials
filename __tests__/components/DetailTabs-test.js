import React from 'react';
import DetailTabs from '../../src/components/DetailTabs';
import { shallow } from '../../src/utils/enzyme';

jest.dontMock('../../src/components/DetailTabs');

describe('DetailsTab component test suite', () => {
  it('contains Tabs and TabPane', () => {
    const docURL = 'https://docs.kony.com/marketplace/login/default.htm';
    const wrappedComponent = shallow(<DetailTabs DocumentationURL={docURL} />);
    expect(wrappedComponent.find('Tabs')).toBeDefined();
    expect(wrappedComponent.find('TabPane')).toBeDefined();
    expect(wrappedComponent.find('Tabs TabPane')).toHaveLength(3);
    expect(wrappedComponent).toMatchSnapshot();
  });

  it('contains asset download button', () => {
    const wrappedComponent = shallow(<DetailTabs />);
    expect(wrappedComponent.find('AssetDownloadButton')).toBeDefined();
    expect(wrappedComponent.find('KonyLike')).toBeDefined();
    expect(wrappedComponent).toMatchSnapshot();
  });
});
