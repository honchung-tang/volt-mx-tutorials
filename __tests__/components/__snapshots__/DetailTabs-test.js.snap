// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsTab component test suite contains Tabs and TabPane 1`] = `
<Tabs
  animated={false}
  className="tabWrapper detailPageTabWrapper"
  defaultActiveKey="1"
  hideAdd={false}
  onChange={[Function]}
  prefixCls="ant-tabs"
  tabBarExtraContent={
    <Row
      className="nocontent"
      gutter={0}
    >
      <Col
        lg={24}
        md={0}
        sm={0}
        xs={0}
      >
        <Connect(AssetDownloadButton)
          DocumentationURL="https://docs.kony.com/marketplace/login/default.htm"
          asset={Object {}}
        />
      </Col>
      <Connect(KonyLike) />
    </Row>
  }
  tabPosition="top"
>
  <TabPane
    key="1"
    placeholder={null}
    tab="Details"
  >
    <DetailsTabContent
      DocumentationURL="https://docs.kony.com/marketplace/login/default.htm"
      asset={Object {}}
    />
  </TabPane>
  <TabPane
    key="2"
    placeholder={null}
    tab="How to"
  >
    <Documentation
      isInIframe={false}
      src="https://docs.kony.com/marketplace/login/default.htm"
    />
  </TabPane>
  <TabPane
    key="3"
    placeholder={null}
    tab="Support"
  >
    <Connect(KonySupport) />
  </TabPane>
</Tabs>
`;

exports[`DetailsTab component test suite contains asset download button 1`] = `
<Tabs
  animated={false}
  className="tabWrapper detailPageTabWrapper"
  defaultActiveKey="1"
  hideAdd={false}
  onChange={[Function]}
  prefixCls="ant-tabs"
  tabBarExtraContent={
    <Row
      className="nocontent"
      gutter={0}
    >
      <Col
        lg={24}
        md={0}
        sm={0}
        xs={0}
      >
        <Connect(AssetDownloadButton)
          DocumentationURL=""
          asset={Object {}}
        />
      </Col>
      <Connect(KonyLike) />
    </Row>
  }
  tabPosition="top"
>
  <TabPane
    key="1"
    placeholder={null}
    tab="Details"
  >
    <DetailsTabContent
      DocumentationURL=""
      asset={Object {}}
    />
  </TabPane>
  <TabPane
    key="3"
    placeholder={null}
    tab="Support"
  >
    <Connect(KonySupport) />
  </TabPane>
</Tabs>
`;
